pkg/reconciler/cache/publishedresources/publishedresources_controller.go

```diff
@@ -0,0 +1,252 @@
+	"github.com/kcp-dev/kcp/pkg/reconciler/committer"
+	cachev1alpha1 "github.com/kcp-dev/kcp/sdk/apis/cache/v1alpha1"
+	kcpclientset "github.com/kcp-dev/kcp/sdk/client/clientset/versioned/cluster"
+	cachev1alpha1client "github.com/kcp-dev/kcp/sdk/client/clientset/versioned/typed/cache/v1alpha1"
+	kcpinformers "github.com/kcp-dev/kcp/sdk/client/informers/externalversions"
+	cacheinformers "github.com/kcp-dev/kcp/sdk/client/informers/externalversions/cache/v1alpha1"
+	cachev1alpha1listers "github.com/kcp-dev/kcp/sdk/client/listers/cache/v1alpha1"
+)
+
+// This Controller starts 1 published resource controller per shard.
+// This is bit wasteful for now. We can optimize this by starting a single controller
+// per logical cluster. But this required stopping and starting the controller when a
+// new published resource is created. This would be fist optimization we need todo as follow-up
+// before removing feature flag and making it enabled by default.
+const (
+	ControllerName = "kcp-published-resources-controller"
+)
+
+// NewController returns a new controller for PublishedResource objects.
+func NewController(
+	kcpClusterClient kcpclientset.ClusterInterface,
+	kcpCacheClient kcpclientset.ClusterInterface,
+	dynamicClient kcpdynamic.ClusterInterface,
+	cacheDynamicClient kcpdynamic.ClusterInterface,
+
+	discoveringDynamicKcpInformers *informer.DiscoveringDynamicSharedInformerFactory,
+	cacheKcpInformers kcpinformers.SharedInformerFactory,
+
+	publishedResourceInformer cacheinformers.PublishedResourceClusterInformer,
+) (*Controller, error) {
+	c := &Controller{
+		queue: workqueue.NewTypedRateLimitingQueueWithConfig(
+			workqueue.DefaultTypedControllerRateLimiter[string](),
+			workqueue.TypedRateLimitingQueueConfig[string]{
+				Name: ControllerName,
+			},
+		),
+		kcpClient:      kcpClusterClient,
+		kcpCacheClient: kcpCacheClient,
+
+		dynamicClient:      dynamicClient,
+		cacheDynamicClient: cacheDynamicClient,
+
+		discoveringDynamicKcpInformers: discoveringDynamicKcpInformers,
+		cacheKcpInformers:              cacheKcpInformers,
+
+		publishedResourceLister:  publishedResourceInformer.Lister(),
+		publishedResourceIndexer: publishedResourceInformer.Informer().GetIndexer(),
+		commit:                   committer.NewCommitter[*cachev1alpha1.PublishedResource, cachev1alpha1client.PublishedResourceInterface, *cachev1alpha1.PublishedResourceSpec, *cachev1alpha1.PublishedResourceStatus](kcpClusterClient.CacheV1alpha1().PublishedResources()),
+
+		conrollerRegistry: newRegistry(),
+	}
+
+	_, _ = publishedResourceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueue(obj) },
+		UpdateFunc: func(_, obj interface{}) { c.enqueue(obj) },
+		DeleteFunc: func(obj interface{}) { c.enqueue(obj) },
+	})
+
+	return c, nil
+}
+
+type publishedResourceResource = committer.Resource[*cachev1alpha1.PublishedResourceSpec, *cachev1alpha1.PublishedResourceStatus]
+
+type Controller struct {
```

Comment:
why is it doing?

Reply:
go doc

Reply:
Nvmd. The constructor has that. I thought there is more. But there is a second reconciler.